File: packages/typescript/temporal/src/redaction.ts
/**
 * Metadata redaction (parity Epic 4, #451) — a port of Python
 * `observability/redaction.py` `RegexPIIRedactor`. `redactMetadata` recursively
 * replaces PII in the string values of a metadata structure (emails / SSNs /
 * Luhn-valid card numbers / phone numbers), skipping configured `excludePaths`
 * (fnmatch-style globs). It is idempotent: a replacement like `[REDACTED_EMAIL]`
 * matches no rule, so re-redacting is a no-op.
 *
 * The redaction is applied to trace metadata by a `TraceWriter` (and exposed here
 * for redacting any free-form metadata before emit).
 */

/** Internal Typeflux/Temporal metadata paths preserved when `preserveTypefluxMetadata` is on. */
export const DEFAULT_EXCLUDED_PATHS: readonly string[] = [
  "temporal.workflow_id",
  "temporal.run_id",
  "temporal.activity_id",
  "temporal.workflow_type",
  "temporal.activity_type",
  "typeflux.workflow.workflow_id",
  "typeflux.workflow.workflow_name",
  "typeflux.runtime_placement.*",
  "typeflux.execution_manifest.*",
  "typeflux.activity.*",
  "typeflux.activity_execution_manifest.*",
  "typeflux.yaml.map_steps.*",
  "typeflux.map.*",
  "typeflux.join.*",
  "typeflux.lifecycle.state",
  "typeflux.lifecycle.current_step",
  "typeflux.lifecycle.completed_units",
  "typeflux.lifecycle.total_units",
  "typeflux.lifecycle.cancellation_requested",
  "typeflux.lifecycle.waiting_checkpoint",
  "typeflux.lifecycle.terminal_status",
  "typeflux.lifecycle.status_event_limit",
  "typeflux.lifecycle.review_after_step",
  "typeflux.provider_controls.*",
  "typeflux.manifest_hash",
  "typeflux.input_schema_hash",
  "typeflux.output_schema_hash",
  "typeflux.activity_execution_manifest_hash",
  // Moderation verdict is a classification (never raw output), preserved as audit evidence.
  "typeflux_moderation.*",
  "langfuse.prompt_config.*",
];

export interface RedactionConfig {
  /** Master switch — `false` returns the value unchanged (default `true`). */
  enabled?: boolean;
  emails?: boolean;
  phones?: boolean;
  ssn?: boolean;
  creditCards?: boolean;
  /** Preserve internal Typeflux/Temporal metadata paths (default `true`). */
  preserveTypefluxMetadata?: boolean;
  /**
   * Extra fnmatch-style glob paths to skip. Paths are relative to the redacted value's
   * ROOT — for trace metadata that is the metadata dict's keys (e.g. `"ticket_id"` or
   * `"temporal.workflow_id"`, NOT `"metadata.ticket_id"`), matching the Python convention
   * and the `DEFAULT_EXCLUDED_PATHS` form.
   */
  excludePaths?: readonly string[];
  /**
   * Custom rules (#188 D188-4) appended AFTER the built-in catalog: the built-ins redact
   * known PII first, then these jurisdiction/domain rules run over the result. `pattern`
   * is the JS `RegExp` dialect (per-edition — the same YAML gives Python `re` on the other
   * side; see docs/privacy.md). Excluded paths still win, checked per node before any rule.
   */
  customRules?: readonly CustomRedactionRule[];
}

/** A custom redaction rule declared in YAML (`observability.redaction.custom_rules`). */
export interface CustomRedactionRule {
  name: string;
  /** JS `RegExp` source; compiled with the global flag so every match is replaced. */
  pattern: string;
  replacement: string;
}

interface RedactionRule {
  pattern: RegExp;
  replacement: string;
  validator?: (candidate: string) => boolean;
}

/** Luhn check used to redact only valid-looking card numbers (matches the Python validator). */
function isLuhnValid(candidate: string): boolean {
  const digits = candidate.replace(/[ -]/g, "");
  if (!/^\d+$/.test(digits) || digits.length < 13 || digits.length > 19) {
    return false;
  }
  let total = 0;
  const reversed = digits.split("").reverse();
  for (let index = 0; index < reversed.length; index += 1) {
    let value = Number(reversed[index]);
    if (index % 2 === 1) {
      value *= 2;
      if (value > 9) {
        value -= 9;
      }
    }
    total += value;
  }
  return total % 10 === 0;
}

function buildRules(config: RedactionConfig): RedactionRule[] {
  const rules: RedactionRule[] = [];
  if (config.emails !== false) {
    rules.push({ pattern: /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/g, replacement: "[REDACTED_EMAIL]" });
  }
  if (config.ssn !== false) {
    rules.push({ pattern: /\b\d{3}-\d{2}-\d{4}\b/g, replacement: "[REDACTED_SSN]" });
  }
  if (config.creditCards !== false) {
    rules.push({
      pattern: /(?<!\d)\d(?:[ -]?\d){12,18}(?!\d)/g,
      replacement: "[REDACTED_CARD]",
      validator: isLuhnValid,
    });
  }
  if (config.phones !== false) {
    rules.push({
      pattern: /(?<!\d)(?:\+1[-.\s]?)?(?:\(\d{3}\)|\d{3}[-.\s])\d{3}[-.\s]\d{4}(?!\d)/g,
      replacement: "[REDACTED_PHONE]",
    });
  }
  for (const rule of config.customRules ?? []) {
    // Global flag so `applyRules`' String.replace replaces every occurrence, matching
    // Python's `pattern.sub`. Compiling here throws SyntaxError on an invalid pattern —
    // the spec/config layer validates first (fail-closed) so this stays a belt-and-braces
    // guard rather than the primary error surface.
    rules.push({ pattern: new RegExp(rule.pattern, "g"), replacement: rule.replacement });
  }
  return rules;
}

/**
 * fnmatchcase-style glob match — `*` matches any run (including dots), `?` one char,
 * everything else literal. A linear two-pointer scan (no regex), so an adversarial
 * `excludePaths` glob can't trigger catastrophic backtracking (parity with Python's
 * `fnmatch`, which compiles to a non-backtracking matcher).
 */
function globMatch(text: string, pattern: string): boolean {
  let t = 0;
  let p = 0;
  let starP = -1;
  let starT = 0;
  while (t < text.length) {
    if (p < pattern.length && (pattern[p] === "?" || pattern[p] === text[t])) {
      t += 1;
      p += 1;
    } else if (p < pattern.length && pattern[p] === "*") {
      starP = p;
      starT = t;
      p += 1;
    } else if (starP !== -1) {
      // Backtrack: let the last `*` absorb one more character.
      p = starP + 1;
      starT += 1;
      t = starT;
    } else {
      return false;
    }
  }
  while (p < pattern.length && pattern[p] === "*") {
    p += 1;
  }
  return p === pattern.length;
}

/**
 * Compiled-rule cache keyed on `RedactionConfig` object IDENTITY. `redactMetadata` is called
 * per trace by the TraceWriter — a single trace fans out to input/output/error/metadata plus
 * one call per observation (~6+5M times for a large trace), and `buildRules` recompiles every
 * custom `RegExp` each time. Identity-keying is sound because configs come from parsed spec
 * objects that are not mutated after construction; a distinct config object (or `undefined`,
 * which we don't cache) rebuilds. A `WeakMap` lets a config be GC'd with no manual eviction.
 * The cached rules are safe to SHARE across calls because `applyRules` only ever calls
 * `String.prototype.replace` (which resets `lastIndex` on the shared /g regexes each time) —
 * it never uses `.test`/`.exec`, so there is no stateful `lastIndex` to leak between calls.
 */
const compiledRulesCache = new WeakMap<RedactionConfig, RedactionRule[]>();

function getRules(config: RedactionConfig | undefined): RedactionRule[] {
  if (config === undefined) {
    return buildRules({});
  }
  const cached = compiledRulesCache.get(config);
  if (cached !== undefined) {
    return cached;
  }
  const rules = buildRules(config);
  compiledRulesCache.set(config, rules);
  return rules;
}

function applyRules(text: string, rules: RedactionRule[]): string {
  let result = text;
  for (const rule of rules) {
    result = result.replace(rule.pattern, (match) =>
      rule.validator === undefined || rule.validator(match) ? rule.replacement : match,
    );
  }
  return result;
}

function redactValue(value: unknown, path: string[], rules: RedactionRule[], excludePaths: readonly string[]): unknown {
  if (path.length > 0) {
    const dotted = path.join(".");
    if (excludePaths.some((pattern) => globMatch(dotted, pattern))) {
      return value;
    }
  }
  if (typeof value === "string") {
    return applyRules(value, rules);
  }
  if (Array.isArray(value)) {
    // Array item paths stay index-less (matches Python — for wildcard excludes like `…activities.*`).
    return value.map((item) => redactValue(item, path, rules, excludePaths));
  }
  if (isPlainObject(value)) {
    const out: Record<string, unknown> = {};
    for (const [key, item] of Object.entries(value)) {
      out[key] = redactValue(item, [...path, key], rules, excludePaths);
    }
    return out;
  }
  // Non-plain objects (Date, Map, Set, URL, class instances) pass through unchanged —
  // rebuilding them from `Object.entries` would lose their type / drop non-enumerable state.
  return value;
}

/** A `{}`-literal / `Object.create(null)` record — NOT a Date/Map/Set/class instance. */
function isPlainObject(value: unknown): value is Record<string, unknown> {
  if (value === null || typeof value !== "object") {
    return false;
  }
  const proto = Object.getPrototypeOf(value);
  return proto === Object.prototype || proto === null;
}

/**
 * Recursively redact PII from the string values of `value`. Returns a redacted copy;
 * the input is not mutated. Disabled (`enabled: false`) returns the value unchanged.
 */
export function redactMetadata<T>(value: T, config?: RedactionConfig): T {
  if (config?.enabled === false) {
    return value;
  }
  // `getRules` memoizes per config-object identity (see `compiledRulesCache`); an omitted
  // config rebuilds without caching a throwaway `{}` on every call.
  const rules = getRules(config);
  const excludePaths =
    config?.preserveTypefluxMetadata !== false
      ? [...DEFAULT_EXCLUDED_PATHS, ...(config?.excludePaths ?? [])]
      : [...(config?.excludePaths ?? [])];
  return redactValue(value, [], rules, excludePaths) as T;
}

File: packages/python/src/typeflux/observability/redaction.py
from __future__ import annotations

import re
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from fnmatch import fnmatchcase
from re import Match, Pattern
from typing import Any, Protocol

from pydantic import BaseModel


class Redactor(Protocol):
    def redact(self, data: Any, **kwargs: Any) -> Any: ...


class NoOpRedactor:
    def redact(self, data: Any, **kwargs: Any) -> Any:
        return data


class _RedactedMetadata(dict[str, Any]):
    pass


def _mark_redacted_metadata(metadata: dict[str, Any]) -> dict[str, Any]:
    if isinstance(metadata, _RedactedMetadata):
        return metadata
    return _RedactedMetadata(metadata)


def _is_redacted_metadata(data: Any) -> bool:
    return isinstance(data, _RedactedMetadata)


@dataclass(frozen=True)
class RegexRedactionRule:
    name: str
    pattern: Pattern[str]
    replacement: str
    validator: Callable[[str], bool] | None = None


DEFAULT_EXCLUDED_PATHS = (
    "temporal.workflow_id",
    "temporal.run_id",
    "temporal.activity_id",
    "temporal.workflow_type",
    "temporal.activity_type",
    "typeflux.workflow.workflow_id",
    "typeflux.workflow.workflow_name",
    "typeflux.runtime_placement.platform",
    "typeflux.runtime_placement.kubernetes.namespace",
    "typeflux.runtime_placement.kubernetes.pod_name",
    "typeflux.runtime_placement.kubernetes.pod_uid",
    "typeflux.runtime_placement.kubernetes.node_name",
    "typeflux.runtime_placement.kubernetes.service_account",
    "typeflux.runtime_placement.kubernetes.deployment_name",
    "typeflux.runtime_placement.kubernetes.worker_name",
    "typeflux.runtime_placement.container_image",
    "typeflux.execution_manifest.workflow_id",
    "typeflux.execution_manifest.temporal_run_id",
    "typeflux.execution_manifest.workflow_contract_hash",
    "typeflux.execution_manifest.manifest_hash",
    "typeflux.execution_manifest.activities.manifest_hash",
    "typeflux.execution_manifest.activities.activity_manifest_hash",
    "typeflux.execution_manifest.activities.input_schema.hash",
    "typeflux.execution_manifest.activities.output_schema.hash",
    "typeflux.execution_manifest.activities.prompt_messages_hash",
    "typeflux.execution_manifest.activities.rendered_messages_hash",
    "typeflux.execution_manifest.activities.provider_params.*",
    "typeflux.execution_manifest.activities.start_to_close_timeout_seconds",
    "typeflux.execution_manifest.activities.definition_source.*",
    "typeflux.execution_manifest.activities.artifact_inputs.*",
    "typeflux.execution_manifest.activities.artifacts.*",
    "typeflux.execution_manifest.map_steps.*",
    "typeflux.execution_manifest.code_provenance.git_sha",
    "typeflux.execution_manifest.code_provenance.git_ref",
    "typeflux.execution_manifest.code_provenance.repo_url",
    "typeflux.yaml.map_steps.*",
    "typeflux.map.*",
    # Keep in sync with _SAFE_LIFECYCLE_FIELDS in typeflux.metadata;
    # arbitrary lifecycle values are not blanket-preserved.
    "typeflux.lifecycle.state",
    "typeflux.lifecycle.current_step",
    "typeflux.lifecycle.completed_units",
    "typeflux.lifecycle.total_units",
    "typeflux.lifecycle.cancellation_requested",
    "typeflux.lifecycle.waiting_checkpoint",
    "typeflux.lifecycle.terminal_status",
    "typeflux.lifecycle.status_event_limit",
    "typeflux.lifecycle.review_after_step",
    "typeflux.provider_controls.*",
    # Moderation verdict is a classification (decision/categories/score/moderator),
    # never raw output, so it is preserved as audit evidence (#158). Recorded as
    # its own top-level key (not nested under typeflux) because a backend
    # update_metadata replaces a top-level key rather than deep-merging it.
    "typeflux_moderation.*",
    "typeflux.activity.manifest_hash",
    "typeflux.activity.input_schema_hash",
    "typeflux.activity.output_schema_hash",
    "typeflux.activity.activity_manifest_hash",
    "typeflux.activity.input_schema.hash",
    "typeflux.activity.output_schema.hash",
    "typeflux.activity.prompt_messages_hash",
    "typeflux.activity.rendered_messages_hash",
    "typeflux.activity.provider_params.*",
    "typeflux.activity.definition_source.*",
    "typeflux.activity.artifact_inputs.*",
    "typeflux.activity.artifacts.*",
    "typeflux.join.*",
    "typeflux.manifest_hash",
    "typeflux.input_schema_hash",
    "typeflux.output_schema_hash",
    "typeflux.activity_execution_manifest_hash",
    "typeflux.activity_execution_manifest.manifest_hash",
    "typeflux.activity_execution_manifest.activity_manifest_hash",
    "typeflux.activity_execution_manifest.input_schema.hash",
    "typeflux.activity_execution_manifest.output_schema.hash",
    "typeflux.activity_execution_manifest.prompt_messages_hash",
    "typeflux.activity_execution_manifest.rendered_messages_hash",
    "typeflux.activity_execution_manifest.provider_params.*",
    "typeflux.activity_execution_manifest.start_to_close_timeout_seconds",
    "typeflux.activity_execution_manifest.definition_source.*",
    "typeflux.activity_execution_manifest.artifact_inputs.*",
    "typeflux.activity_execution_manifest.artifacts.*",
    "typeflux.registry.langfuse.prompt_config.typeflux.contracts.input_schema_hash",
    "typeflux.registry.langfuse.prompt_config.typeflux.contracts.output_schema_hash",
    "langfuse.prompt_config.typeflux.contracts.input_schema_hash",
    "langfuse.prompt_config.typeflux.contracts.output_schema_hash",
)


class RegexPIIRedactor:
    def __init__(
        self,
        rules: tuple[RegexRedactionRule, ...],
        *,
        exclude_paths: Sequence[str] = (),
    ) -> None:
        self.rules = rules
        self.exclude_paths = tuple(exclude_paths)

    @classmethod
    def default(
        cls,
        *,
        emails: bool = True,
        phones: bool = True,
        ssn: bool = True,
        credit_cards: bool = True,
        exclude_paths: Sequence[str] = (),
        preserve_typeflux_metadata: bool = True,
        custom_rules: Sequence[RegexRedactionRule] = (),
    ) -> RegexPIIRedactor:
        rules: list[RegexRedactionRule] = []
        if emails:
            rules.append(
                RegexRedactionRule(
                    name="email",
                    pattern=re.compile(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b"),
                    replacement="[REDACTED_EMAIL]",
                )
            )
        if ssn:
            rules.append(
                RegexRedactionRule(
                    name="ssn",
                    pattern=re.compile(r"\b\d{3}-\d{2}-\d{4}\b"),
                    replacement="[REDACTED_SSN]",
                )
            )
        if credit_cards:
            rules.append(
                RegexRedactionRule(
                    name="credit_card",
                    pattern=re.compile(r"(?<!\d)\d(?:[ -]?\d){12,18}(?!\d)"),
                    replacement="[REDACTED_CARD]",
                    validator=_is_luhn_valid_candidate,
                )
            )
        if phones:
            rules.append(
                RegexRedactionRule(
                    name="phone",
                    pattern=re.compile(
                        r"(?<!\d)(?:\+1[-.\s]?)?(?:\(\d{3}\)|\d{3}[-.\s])\d{3}[-.\s]\d{4}(?!\d)"
                    ),
                    replacement="[REDACTED_PHONE]",
                )
            )
        # Custom rules (#188 D188-4) append AFTER the built-in catalog: the built-ins
        # redact known PII first, then jurisdiction/domain rules run over the result.
        # DEFAULT_EXCLUDED_PATHS still wins — exclusion is checked per node in
        # `_redact_value` BEFORE any rule (built-in or custom) touches a string.
        rules.extend(custom_rules)
        resolved_exclude_paths = (
            (*DEFAULT_EXCLUDED_PATHS, *exclude_paths)
            if preserve_typeflux_metadata
            else tuple(exclude_paths)
        )
        return cls(tuple(rules), exclude_paths=resolved_exclude_paths)

    def redact(self, data: Any, **kwargs: Any) -> Any:
        return self._redact_value(data, path=())

    def _redact_value(self, value: Any, *, path: tuple[str, ...]) -> Any:
        if self._is_excluded(path):
            return value
        if isinstance(value, str):
            return self._redact_string(value)
        if isinstance(value, BaseModel):
            return self._redact_value(value.model_dump(mode="python"), path=path)
        if isinstance(value, dict):
            return {
                key: self._redact_value(item, path=(*path, str(key))) for key, item in value.items()
            }
        # Sequence item paths intentionally remain index-less for compatibility with
        # existing wildcard excludes such as typeflux.execution_manifest.activities.*.
        if isinstance(value, list):
            return [self._redact_value(item, path=path) for item in value]
        if isinstance(value, tuple):
            return tuple(self._redact_value(item, path=path) for item in value)
        if isinstance(value, set):
            return {self._redact_value(item, path=path) for item in value}
        if isinstance(value, frozenset):
            return frozenset(self._redact_value(item, path=path) for item in value)
        return value

    def _redact_string(self, value: str) -> str:
        redacted = value
        for rule in self.rules:
            redacted = rule.pattern.sub(_replacement(rule), redacted)
        return redacted

    def _is_excluded(self, path: tuple[str, ...]) -> bool:
        if not path:
            return False
        dotted_path = ".".join(path)
        return any(fnmatchcase(dotted_path, pattern) for pattern in self.exclude_paths)


def _replacement(rule: RegexRedactionRule) -> Callable[[Match[str]], str]:
    validator = rule.validator
    if validator is None:
        # LITERAL replacement. Return a function so `re.sub` treats `rule.replacement` as
        # a literal string and never interprets `\1` / `\g<name>` / backslash escapes in
        # it. A plain string repl (the previous behavior) would, for a YAML custom rule
        # like `replacement: '[CASE \1]'`, re-inject the CAPTURED PII into the "redacted"
        # output, and a stray backslash (e.g. `\C`) would raise re.error AT TRACE EMISSION.
        # Built-in rules contain no backslashes, so their behavior is unchanged. We do NOT
        # re.escape the replacement — that would emit doubled backslashes into the output.
        # This also matches the TS edition, whose `applyRules` returns the replacement from
        # a replacer FUNCTION (never `$`-interpreted), so identical YAML redacts identically.
        return lambda _match: rule.replacement

    def replace(match: Match[str]) -> str:
        candidate = match.group(0)
        return rule.replacement if validator(candidate) else candidate

    return replace


def _is_luhn_valid_candidate(candidate: str) -> bool:
    digits = candidate.replace(" ", "").replace("-", "")
    if not digits.isdigit() or not 13 <= len(digits) <= 19:
        return False
    return _luhn_checksum(digits) == 0


def _luhn_checksum(digits: str) -> int:
    total = 0
    reverse_digits = digits[::-1]
    for index, char in enumerate(reverse_digits):
        value = int(char)
        if index % 2 == 1:
            value *= 2
            if value > 9:
                value -= 9
        total += value
    return total % 10


__all__ = [
    "DEFAULT_EXCLUDED_PATHS",
    "NoOpRedactor",
    "Redactor",
    "RegexPIIRedactor",
    "RegexRedactionRule",
]
